RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETMemberMetricCalculator.cs

```diff
@@ -121,5 +137,404 @@ private int CountOccurrences(string text, string pattern)
             }
             return count;
         }
+
+        private int CountTryCatchBlocks(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<TryStatementSyntax>().Count();
+        }
+
+        private int CountLoops(MemberDeclarationSyntax method)
+        {
+            int count = method.DescendantNodes().OfType<ForEachStatementSyntax>().Count();
+            count += method.DescendantNodes().OfType<ForStatementSyntax>().Count();
+            count += method.DescendantNodes().OfType<WhileStatementSyntax>().Count();
+            count += method.DescendantNodes().OfType<DoStatementSyntax>().Count();
+            return count;
+        }
+
+        private int CountReturnStatements(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<ReturnStatementSyntax>().Count();
+        }
+
+        private int CountComparisonOperators(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<BinaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.EqualsExpression) || 
```

Comment:
Would this also count https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-coalescing-operator ?

If I'm not mistaking, this would map to "CoalesceExpression" but you should test it (by writing a test/expanding test data).

Reply:
The current implementation of the method does not include coalesce operators (?? and ?? =). CoalesceExpression would catch this example "b = a ?? -1;", and CoalesceAssignmentExpression would catch this example "b ?? = 2;". Should I include this?